Basic support for viaIR

fvictorio opened the pull request:
Closes #2724.
Closes #1533

This PR is (almost) the same as #3205, but with a better commits structure. I recommend reading the commits one by one when reviewing this.

fvictorio commented:
Results of using this in the OpenZeppelin repo, enabling viaIR and running their tests:

- If you use solc 0.8.18 (nightly) and set `":"` as the optimizer steps, all tests pass.
- If you use solc 0.8.17 and set "" as the optimizer steps, there are 34 failures, but all of them are caused by https://github.com/ethereum/solidity/issues/13786.
- If you use solc 0.8.17 with the default optimizer steps, there are 44 failures. All of them are caused by Hardhat returning `"Transaction reverted and Hardhat couldn't infer the reason"` instead of the expected error. Since this scenario is using both the new codegen and optimizations enabled, I feel this is acceptable. 

  We _could_ improve these results, though, if we had a better implementation of the `ErrorInferrer` that _always_ returns the correct top stack trace entry when there is return data. I tried to do this but I wasn't able to get it right.

ekpyron commented:
It may be worth a try for you to use ``"u"`` as optimizer step sequence for 0.8.17 - that should be powerful enough to eliminate the link references, but weak enough not to interfere with anything else.

fvictorio commented:
Indeed! All OZ tests pass with `optimizerSteps: "u"`. Thanks a lot, @ekpyron!

fvictorio commented on packages/hardhat-core/src/internal/cli/cli.ts:
```diff
@@ -104,6 +104,23 @@ async function suggestInstallingHardhatVscode() {
+function showViaIRWarning(resolvedConfig: HardhatConfig) {
+  const viaIREnabled = resolvedConfig.solidity.compilers.some(
+    (compiler) => compiler.settings?.viaIR === true
+  );
```
We should check the overrides too.

fvictorio commented on packages/hardhat-core/test/internal/hardhat-network/stack-traces/compilation.ts:
```diff
@@ -33,41 +31,70 @@ function getSolcSourceFileMapping(sources: string[]): SolcSourceFileToContents {
+    semver.gte(compilerOptions.solidityVersion, "0.8.18") && isViaIR
+      ? {
+          yulDetails: {
+            optimizerSteps: ":",
```
Use `"u"` and remove `0.8.18` check

alcuadrado commented:
I think this commit doesn't work well with overloaded functions 1249a530410e80c11d8ee12c997b7ccb4b6a08f0, apart from that it LGTM

alcuadrado commented:
Should the guide recommend  using `"u"` for solc < 0.8.18 and `":"` otherwise?

alcuadrado commented:
>  it depends on your heuristics if "u" or "u:" works better (the latter has fewer optimizations than the former), but it's probably a good idea to keep including the u in any case.

We'll try it out and let you know.

> We're by the way considering [ethereum/solidity#13972](https://github.com/ethereum/solidity/issues/13972) for 0.8.20 - that may make this smoother.

That's great!

btw, we are planning to release this next week.

If anyone want's to test it, I just published a build under `hardhat@viair`

fvictorio commented:
> I think this commit doesn't work well with overloaded functions

Ugh, I'm glad you caught that, thank you.

I fixed it by passing all the function abis that match the name, and then checking if any of those matches the selector of the function node. I also added a test that triggered the problem. See https://github.com/NomicFoundation/hardhat/pull/3415/commits/dea67989e2d9f9ada6b2b225f3465b00095a129e.
